File: scc/scssgo/graph.go
package main

import (
	"fmt"
)

// Edge represents a directed edge u -> v.
type Edge struct {
	U, V int
}

type Graph struct {
	V      int
	Adj    [][]int
	RevAdj [][]int
	Verbose bool
}

func NewGraph(v int) *Graph {
	g := &Graph{
		V:      v,
		Adj:    make([][]int, v),
		RevAdj: make([][]int, v),
		Verbose: false,
	}
	for i := 0; i < v; i++ {
		g.Adj[i] = []int{}
		g.RevAdj[i] = []int{}
	}
	return g
}

func (g *Graph) AddEdge(v, w int) {
	g.Adj[v] = append(g.Adj[v], w)
	g.RevAdj[w] = append(g.RevAdj[w], v) // reverse graph for later use
}

// ---------- Tarjan’s SCC (O(V+E)) ----------

func (g *Graph) tarjanDFS(
	u int,
	disc []int,
	low []int,
	stack *[]int,
	inStack []bool,
	timeRef *int,
	sccList *[][]int,
) {
	*timeRef += 1
	disc[u] = *timeRef
	low[u] = *timeRef
	*stack = append(*stack, u)
	inStack[u] = true

	for _, v := range g.Adj[u] {
		if disc[v] == -1 {
			g.tarjanDFS(v, disc, low, stack, inStack, timeRef, sccList)
			if low[v] < low[u] {
				low[u] = low[v]
			}
		} else if inStack[v] {
			if disc[v] < low[u] {
				low[u] = disc[v]
			}
		}
	}

	if low[u] == disc[u] {
		var scc []int
		for {
			w := (*stack)[len(*stack)-1]
			*stack = (*stack)[:len(*stack)-1]
			inStack[w] = false
			scc = append(scc, w)
			if w == u {
				break
			}
		}
		*sccList = append(*sccList, scc)
	}
}

func (g *Graph) FindSCCs() [][]int {
	disc := make([]int, g.V)
	low := make([]int, g.V)
	inStack := make([]bool, g.V)
	stack := make([]int, 0, g.V)
	sccList := make([][]int, 0)

	for i := 0; i < g.V; i++ {
		disc[i] = -1
		low[i] = -1
	}
	timeRef := 0

	for i := 0; i < g.V; i++ {
		if disc[i] == -1 {
			g.tarjanDFS(i, disc, low, &stack, inStack, &timeRef, &sccList)
		}
	}
	return sccList
}

// ---------- Minimal SCC Edge Reduction ----------

func (g *Graph) MinimizeEdgesInSCC(scc []int) []Edge {
	nodes := make(map[int]bool, len(scc))
	for _, n := range scc {
		nodes[n] = true
	}

	// Step 1: forward spanning tree using DFS
	forwardTree := g.buildSpanningTree(scc[0], g.Adj, nodes)

	// Step 2: reverse spanning tree using DFS on reversed graph
	reverseTree := g.buildSpanningTree(scc[0], g.RevAdj, nodes)

	// Step 3: merge both trees (each edge appears at most twice)
	out := make([]Edge, 0, len(forwardTree)+len(reverseTree))
	for e := range forwardTree {
		out = append(out, e)
	}
	for e := range reverseTree {
		out = append(out, e)
	}
	return out
}

func (g *Graph) buildSpanningTree(start int, graph [][]int, nodes map[int]bool) map[Edge]struct{} {
	spanning := make(map[Edge]struct{})
	visited := make(map[int]bool, len(nodes))
	stack := []int{start}
	visited[start] = true

	for len(stack) > 0 {
		// pop
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		for _, nb := range graph[node] {
			if nodes[nb] && !visited[nb] {
				spanning[Edge{node, nb}] = struct{}{}
				visited[nb] = true
				stack = append(stack, nb)
			}
		}
	}
	return spanning
}

func (g *Graph) ReduceEdges() []Edge {
	sccs := g.FindSCCs()
	if g.Verbose {
		fmt.Printf("Found %d SCC(s).\n", len(sccs))
	}

	reduced := make([]Edge, 0)
	for _, scc := range sccs {
		minEdges := g.MinimizeEdgesInSCC(scc)
		reduced = append(reduced, minEdges...)
	}

	if g.Verbose {
		fmt.Printf("Reduced SCC edges: %d\n", len(reduced))
	}
	return reduced
}

File: scc/python/graph.py
from typing import List, Set, Tuple

class Graph:
    def __init__(self, v: int, verbose: bool = False) -> None:
        self.V: int = v
        self.adj: List[List[int]] = [[] for _ in range(v)]
        self.rev_adj: List[List[int]] = [[] for _ in range(v)]
        self.verbose = verbose

    def add_edge(self, v: int, w: int) -> None:
        self.adj[v].append(w)
        self.rev_adj[w].append(v)
    
    def tarjan_dfs(self,
        u: int,
        disc: List[int],
        low: List[int],
        stack: List[int],
        in_stack: List[bool],
        time_ref: List[int],
        scc_list: List[List[int]],
    ) -> None:
        time_ref[0] += 1
        disc[u] = low[u] = time_ref[0]
        stack.append(u)
        in_stack[u] = True                

        for v in self.adj[u]:
            if disc[v] == -1:
                self.tarjan_dfs(v, disc, low, stack, in_stack, time_ref, scc_list)
                low[u] = min(low[u], low[v])
            elif in_stack[v]:
                low[u] = min(low[u], disc[v])
        
        if low[u] == disc[u]:
            scc: List[int] = []
            while True:
                w = stack.pop()
                in_stack[w] = False
                scc.append(w)
                if w == u:
                    break
            scc_list.append(scc)

    
    def find_sccs(self) -> List[List[int]]:
        disc = [-1] * self.V
        low = [-1] * self.V
        in_stack = [False] * self.V
        stack: List[int] = []
        scc_list: List[List[int]] = []
        time_ref = [0]  # acts like 'ref int time'

        for i in range(self.V):
            if disc[i] == -1:
                self.tarjan_dfs(i, disc, low, stack, in_stack, time_ref, scc_list)

        return scc_list

    # Minimal SCC Edge Reduction (O(V + E))
    def minimize_edges_in_scc(self, scc: List[int]) -> List[Tuple[int, int]]:
        nodes: Set[int] = set(scc)
        essential_edges: List[Tuple[int, int]] = []

        # Step 1: forward spanning tree using DFS
        forward_tree = self.build_spanning_tree(scc[0], self.adj, nodes)

        # Step 2: reverse spanning tree using DFS on reversed graph
        reverse_tree = self.build_spanning_tree(scc[0], self.rev_adj, nodes)

        # Step 3: merge both trees (each edge appears at most twice)
        essential_edges.extend(forward_tree)
        essential_edges.extend(reverse_tree)

        return essential_edges

    def build_spanning_tree(
        self, start: int, graph: List[List[int]], nodes: Set[int]
    ) -> Set[Tuple[int, int]]:
        spanning_tree: Set[Tuple[int, int]] = set()
        visited: Set[int] = set()
        stack: List[int] = [start]
        visited.add(start)
        
        while stack:
            node = stack.pop()
            for neighbor in graph[node]:
                if neighbor in nodes and neighbor not in visited:
                    spanning_tree.add((node, neighbor))
                    visited.add(neighbor)
                    stack.append(neighbor)
        
        return spanning_tree

    def reduce_edges(self) -> List[Tuple[int, int]]:
        SCCs = self.find_sccs()
        if self.verbose:
            print(f"Found {len(SCCs)} SCC(s).")

        reduced_edges: List[Tuple[int, int]] = []
        for scc in SCCs:
            min_edges = self.minimize_edges_in_scc(scc)
            reduced_edges.extend(min_edges)

        if self.verbose:
            print(f"Reduced SCC edges: {len(reduced_edges)}")
        return reduced_edges

Graph.reduce_edges